crates/mpz-garble-core/src/generator.rs

```diff
@@ -140,111 +126,254 @@ impl Generator {
+    /// * `inputs` - The input values to the circuit.
+    pub fn generate_batched<'a>(
+        &'a mut self,
+        circ: &'a Circuit,
+        delta: Delta,
+        inputs: Vec<EncodedValue<state::Full>>,
+    ) -> Result<EncryptedGateBatchIter<'_, std::slice::Iter<'_, Gate>>, GeneratorError> {
+        self.generate(circ, delta, inputs)
+            .map(EncryptedGateBatchIter)
+    }
+}
+
+/// Iterator over encrypted gates of a garbled circuit.
+pub struct EncryptedGateIter<'a, I> {
+    /// Cipher to use to encrypt the gates
+    cipher: &'static FixedKeyAes,
+    /// Global offset.
+    delta: Delta,
+    /// Buffer for the 0-bit labels.
+    labels: &'a mut [Label],
+    /// Iterator over the gates.
+    gates: I,
+    /// Circuit outputs.
+    outputs: &'a [BinaryRepr],
+    /// Current gate id.
+    gid: usize,
+    /// Hasher to use to hash the encrypted gates
+    hasher: Option<Hasher>,
+    /// Number of AND gates generated.
+    counter: usize,
+    /// Number of AND gates in the circuit.
+    and_count: usize,
+    /// Whether the entire circuit has been garbled.
+    complete: bool,
+}
+
+impl<'a, I> fmt::Debug for EncryptedGateIter<'a, I> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "EncryptedGateIter {{ .. }}")
+    }
+}
+
+impl<'a, I> EncryptedGateIter<'a, I>
+where
+    I: Iterator<Item = &'a Gate>,
+{
+    fn new(
+        delta: Delta,
+        gates: I,
+        outputs: &'a [BinaryRepr],
+        labels: &'a mut [Label],
+        and_count: usize,
+    ) -> Self {
+        Self {
             cipher: &(*FIXED_KEY_AES),
-            circ,
             delta,
-            low_labels,
-            pos: 0,
+            gates,
+            outputs,
+            labels,
             gid: 1,
-            hasher,
-        })
+            hasher: None,
+            counter: 0,
+            and_count,
+            complete: false,
+        }
+    }
+
+    /// Enables hashing of the encrypted gates.
+    pub fn enable_hasher(&mut self) {
+        self.hasher = Some(Hasher::new());
     }
 
-    /// Returns whether the generator has finished generating the circuit.
-    pub fn is_complete(&self) -> bool {
-        self.pos >= self.circ.gates().len()
+    /// Returns `true` if the generator has encrypted gates to send.
+    #[inline]
+    pub fn has_gates(&self) -> bool {
+        self.counter != self.and_count
     }
 
-    /// Returns the encoded outputs of the circuit.
-    pub fn outputs(&self) -> Result<Vec<EncodedValue<state::Full>>, GeneratorError> {
-        if !self.is_complete() {
+    /// Returns the encoded outputs of the circuit, and the hash of the encrypted gates if present.
+    pub fn finish(mut self) -> Result<GeneratorOutput, GeneratorError> {
+        if self.has_gates() {
             return Err(GeneratorError::NotFinished);
         }
 
-        Ok(self
-            .circ
-            .outputs()
+        // Finish computing any "free" gates.
```

Comment:
Like above, this is a special case for 0 AND gates.
```suggestion
        // If there were 0 AND gates in the circuit, we compute any "free" gates now.
```

Reply:
@th4s  answered this above

Reply:
My suggested comment is still valid, ill explain on the call today.